### DISCUSSION TITLE
useHydrateAtoms cause global Layout component to remount on route change in Next.js

### DISCUSSION THREAD

**Discussion by AlexTechNoir** [pmndrs/jotai#2035] [Q&A] [answered]
Good day!

I have a Next.js app with global Layout component and useEffect inside it, that fetches global data. I noticed that because of `useHydrateAtoms`, Layout component remounts on every route change and thus, global data gets refetched which affects UI (skeletons, loaders) inside Layout component. When I comment `HydrateAtoms` wrapper out, everything's okay

Here's the [reproduction example](https://main--soft-cajeta-1d2d61.netlify.app/) with [repo](https://github.com/AlexTechNoir/jotai-use-hydrate-atoms-example)

What can I do to resolve this?

**Comment by dai-shi** [comment]
> When I comment `HydrateAtoms` wrapper out, everything's okay

That sounds unexpected to me.
What if you just comment out `useHydrateAtoms()` line?

Can anyone investigate how it happens?

**Reply by AlexTechNoir** [reply]
> What if you just comment out useHydrateAtoms() line?

Tried, nothing's changed, Layout's still remounting on route change

> Can anyone investigate how it happens?

I've set up an [example](https://stackblitz.com/edit/github-ycqylf?file=pages%2F_app.js) - you can try it yourself - comment out `HydrateAtoms` wrapper or `useHydrateAtoms()` line, save changes (Ctrl+S), open in new tab (right top button), then open console and click on links

**Reply by dai-shi** [reply]
> Tried, nothing's changed, Layout's still remounting on route change

Then, what causes the remounting?

> comment out `HydrateAtoms` wrapper or `useHydrateAtoms()` line

So, do you mean commenting out `useHydrateAtoms()` fixes the behavior? I'm confused.

> Can anyone investigate how it happens?

I was simply asking someone to debug it.

**Reply by AlexTechNoir** [reply]
Remounting is caused by `HydrateAtoms` wrapper, it seems, not `useHydrateAtoms()` hook itself, because when I delete the wrapper component from the code, Layout is not remounting on route change, as it should. Deleting [this line](https://github.com/AlexTechNoir/jotai-use-hydrate-atoms-example/blob/main/pages/_app.js#L13) doesn't fix the problem

**Reply by dai-shi** [reply, marked as answer]
Oh, I know why. You define `HydrateAtoms` in the render function. You must put it outside of `App`.

**Reply by AlexTechNoir** [reply]
Thanks for your time, it works now!
